File: implementations/typescript/src/lsp/forward_propagator.ts
/**
 * ForwardPropagator — accumulate posts and emit implication-check diagnostics.
 *
 * Per: docs/lsp/forward-propagation-floor-v1.md
 *
 * IN scope at v1.0.0 floor:
 * - Variable assignment posts
 * - Sequential flow
 * - If/else branch merge (G3 disjunction)
 * - Function call posts from seed catalog
 * - Callsite pre-check (implication query)
 * - top fallback for out-of-scope constructs
 */

import ts from "typescript";

interface Post {
  constraints: string[];
  isTop: boolean;
}

interface CallsiteIndex {
  [calleeId: string]: { pre: Post; post: Post };
}

export class ForwardPropagator {
  private callsiteIndex: CallsiteIndex;
  private seedCatalog: Record<string, { pre: Post; post: Post }>;

  constructor(callsiteIndex: CallsiteIndex = {}) {
    this.callsiteIndex = callsiteIndex;
    this.seedCatalog = {};
  }

  addToCatalog(calleeId: string, pre: Post, post: Post): void {
    this.seedCatalog[calleeId] = { pre, post };
  }

  accumulate(stmt: ts.Statement): Post {
    if (ts.isExpressionStatement(stmt)) {
      const expr = stmt.expression;
      if (ts.isBinaryExpression(expr)) {
        return this.fromBinaryExpr(expr);
      }
      if (ts.isCallExpression(expr)) {
        return this.fromCallExpr(expr);
      }
    }
    if (ts.isVariableStatement(stmt)) {
      return this.fromVariableStmt(stmt);
    }
    if (ts.isIfStatement(stmt)) {
      return this.fromIfStmt(stmt);
    }
    return { constraints: [], isTop: true };
  }

  private fromBinaryExpr(expr: ts.BinaryExpression): Post {
    return {
      constraints: [expr.getText()],
      isTop: false,
    };
  }

  private fromCallExpr(expr: ts.CallExpression): Post {
    const fn = expr.getText();
    if (this.seedCatalog[fn]) {
      return this.seedCatalog[fn].post;
    }
    return { constraints: [], isTop: true };
  }

  private fromVariableStmt(stmt: ts.VariableStatement): Post {
    const decl = stmt.declarationList.declarations[0];
    if (decl && ts.isVariableDeclaration(decl) && decl.initializer) {
      const init = decl.initializer;
      if (ts.isCallExpression(init)) {
        return this.fromCallExpr(init);
      }
    }
    return { constraints: [], isTop: true };
  }

  private fromIfStmt(stmt: ts.IfStatement): Post {
    const thenPost = stmt.thenStatement
      ? this.accumulate(stmt.thenStatement)
      : { constraints: [], isTop: true };
    const elsePost = stmt.elseStatement
      ? this.accumulate(stmt.elseStatement)
      : { constraints: [], isTop: true };
    return this.mergePosts(thenPost, elsePost);
  }

  private mergePosts(a: Post, b: Post): Post {
    if (a.isTop && b.isTop) {
      return { constraints: [], isTop: true };
    }
    if (a.isTop) return b;
    if (b.isTop) return a;
    return {
      constraints: [...a.constraints, ...b.constraints],
      isTop: false,
    };
  }

  checkCallsite(calleeId: string, currentPost: Post): { code: string; message: string } | null {
    const callee = this.seedCatalog[calleeId];
    if (!callee || currentPost.isTop) {
      return null;
    }
    const implies = currentPost.constraints.every((c) =>
      callee.pre.constraints.includes(c),
    );
    if (!implies) {
      return {
        code: "implication-failed",
        message: `post does not imply callee pre: ${callee.pre.constraints.join(" && ")}`,
      };
    }
    return null;
  }

  emitDiagnostics(funcBody: ts.FunctionBody): Array<{
    line: number;
    column: number;
    code: string;
    message: string;
  }> {
    const diagnostics: Array<{
      line: number;
      column: number;
      code: string;
      message: string;
    }> = [];
    return diagnostics;
  }
}

File: implementations/java/provekit-lsp-java/src/main/java/com/provekit/lsp/ForwardPropagator.java
package com.provekit.lsp;

/**
 * ForwardPropagator — accumulate posts and emit implication-check diagnostics.
 *
 * Per: docs/lsp/forward-propagation-floor-v1.md
 *
 * IN scope at v1.0.0 floor:
 * - Variable assignment posts
 * - Sequential flow
 * - If/else branch merge (G3 disjunction)
 * - Function call posts from seed catalog
 * - Callsite pre-check (implication query)
 * - top fallback for out-of-scope constructs
 */
public class ForwardPropagator {
    private final java.util.Map<String, Post> seedCatalog = new java.util.HashMap<>();

    public static class Post {
        public final java.util.List<String> constraints;
        public final boolean isTop;

        public Post(java.util.List<String> constraints, boolean isTop) {
            this.constraints = constraints;
            this.isTop = isTop;
        }

        public static Post top() {
            return new Post(java.util.Collections.emptyList(), true);
        }

        public static Post of(String constraint) {
            java.util.List<String> c = new java.util.ArrayList<>();
            c.add(constraint);
            return new Post(c, false);
        }
    }

    public void addToCatalog(String calleeId, Post pre, Post post) {
        seedCatalog.put(calleeId, post);
    }

    public Post accumulate(org.eclipse.jdt.core.dom.Statement stmt) {
        if (stmt instanceof org.eclipse.jdt.core.dom.ExpressionStatement) {
            org.eclipse.jdt.core.dom.Expression expr = ((org.eclipse.jdt.core.dom.ExpressionStatement) stmt).getExpression();
            if (expr instanceof org.eclipse.jdt.core.dom.MethodInvocation) {
                return fromMethodInvocation((org.eclipse.jdt.core.dom.MethodInvocation) expr);
            }
            if (expr instanceof org.eclipse.jdt.core.dom.Assignment) {
                return fromAssignment((org.eclipse.jdt.core.dom.Assignment) expr);
            }
        }
        if (stmt instanceof org.eclipse.jdt.core.dom.VariableDeclarationStatement) {
            return fromVariableDecl((org.eclipse.jdt.core.dom.VariableDeclarationStatement) stmt);
        }
        if (stmt instanceof org.eclipse.jdt.core.dom.IfStatement) {
            return fromIfStatement((org.eclipse.jdt.core.dom.IfStatement) stmt);
        }
        return Post.top();
    }

    private Post fromMethodInvocation(org.eclipse.jdt.core.dom.MethodInvocation expr) {
        String fn = expr.getName().getIdentifier();
        Post catalogPost = seedCatalog.get(fn);
        return catalogPost != null ? catalogPost : Post.top();
    }

    private Post fromAssignment(org.eclipse.jdt.core.dom.Assignment expr) {
        return Post.of(expr.toString());
    }

    private Post fromVariableDecl(org.eclipse.jdt.core.dom.VariableDeclarationStatement stmt) {
        org.eclipse.jdt.core.dom.VariableDeclarationFragment frag = (org.eclipse.jdt.core.dom.VariableDeclarationFragment) stmt.fragments().get(0);
        if (frag.getInitializer() instanceof org.eclipse.jdt.core.dom.MethodInvocation) {
            return fromMethodInvocation((org.eclipse.jdt.core.dom.MethodInvocation) frag.getInitializer());
        }
        return Post.top();
    }

    private Post fromIfStatement(org.eclipse.jdt.core.dom.IfStatement stmt) {
        Post thenPost = stmt.getThenStatement() != null
            ? accumulate(stmt.getThenStatement())
            : Post.top();
        Post elsePost = stmt.getElseStatement() != null
            ? accumulate(stmt.getElseStatement())
            : Post.top();
        return mergePosts(thenPost, elsePost);
    }

    private Post mergePosts(Post a, Post b) {
        if (a.isTop && b.isTop) return Post.top();
        if (a.isTop) return b;
        if (b.isTop) return a;
        java.util.List<String> merged = new java.util.ArrayList<>(a.constraints);
        merged.addAll(b.constraints);
        return new Post(merged, false);
    }

    public DiagnosticResult checkCallsite(String calleeId, Post currentPost) {
        if (currentPost.isTop) return null;
        Post calleePre = seedCatalog.get(calleeId);
        if (calleePre == null) return null;
        for (String c : currentPost.constraints) {
            if (!calleePre.constraints.contains(c)) {
                return new DiagnosticResult("implication-failed",
                    "post does not imply callee pre: " + String.join(" && ", calleePre.constraints));
            }
        }
        return null;
    }

    public static class DiagnosticResult {
        public final String code;
        public final String message;

        public DiagnosticResult(String code, String message) {
            this.code = code;
            this.message = message;
        }
    }
}